### PR TITLE
Add *.razor and *.cshtml additional files to the compilation

### LINKED ISSUE #9007: Add *.razor and *.cshtml additional files to the compilation
Similar to https://github.com/SonarSource/sonar-dotnet-autoscan/issues/690.

- Pass ContentFiles to the SolutionBuilder via [AddAdditionalDocument](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.project.addadditionaldocument?view=roslyn-dotnet-4.7.0&devlangs=csharp&f1url=%3FappId%3DDev17IDEF1%26l%3DEN-US%26k%3Dk(Microsoft.CodeAnalysis.Project.AddAdditionalDocument)%3Bk(DevLang-csharp)%26rd%3Dtrue).
- Inside the constructor, filter them to *.razor and *.cshtml, and store only those in a field. It will be used by EditorConfigGenerator.

### PR DESCRIPTION
**Pull request by sebastien-marichal**
Fixes #9007

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.TestFramework/Build/ProjectBuilder.cs [hunk 1]
```diff
@@ -78,11 +81,21 @@ public ProjectBuilder AddDocument(string path)
...
-            : throw new ArgumentException($"The file extension '{fileInfo.Extension}' does not match the project language '{project.Language}' nor Razor.", nameof(path));
+
+        if (!IsExtensionOfSupportedType(fileInfo.Extension))
+        {
+            throw new ArgumentException($"The file extension '{fileInfo.Extension}' does not match the project language '{project.Language}' nor Razor.", nameof(path));
+        }
+
+        return IsRazorDocument(fileInfo.Extension)
+            ? AddAdditionalDocument(project, relativePathFromTestCases, File.ReadAllText(fileInfo.FullName, Encoding.UTF8))
+            : AddDocument(project, relativePathFromTestCases, File.ReadAllText(fileInfo.FullName, Encoding.UTF8));
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/tests/SonarAnalyzer.TestFramework/Build/ProjectBuilder.cs:92]: This is a bit confusing to me, I much prefer the implementation in Autoscan for 2 reasons:
- the `AddDocument` is calling `AddDocument`, but sometimes for razor content files `AddAdditionalDocument`, which is different (even tho the names are similar).
- with this approach, we are required to filter twice the paths for the razor/cshtml files to load them in the EditorConfigGenerator.

I would suggest adding two dedicated methods `AddAdditionalDocuments` and `AddAdditionalDocument`, extracting the logic in some common functionality and keeping the invocations separate. Then in the Verifier.cs `CreateProject` we will filter the files once and call `AddAdditionalDocuments` for razor files (stored in some private field) and then `AddDocuments` for everything else. WDYT?

File: analyzers/tests/SonarAnalyzer.TestFramework.Test/Build/ProjectBuilderTest.cs [hunk 2]
```diff
@@ -66,27 +66,33 @@ public void AddDocument_InvalidExtension()
...
     [TestMethod]
-    public void AddDocument_VbnetDoesntSupportCshtmlFiles() =>
-        EmptyVB.Invoking(x => x.AddDocument(@"TestCases\ProjectBuilder.AddDocument.cshtml").FindDocument("ProjectBuilder.AddDocument.cshtml")).Should().Throw<ArgumentException>();
+    public void AddAdditionalDocument_VbnetDoesntSupportCshtmlFiles() =>
+        EmptyVB.Invoking(x => x.AddAdditionalDocument(@"TestCases\ProjectBuilder.AddDocument.cshtml").FindAdditionalDocument("ProjectBuilder.AddDocument.cshtml")).Should().Throw<ArgumentException>();
 
     [TestMethod]
-    public void AddDocument_CsharpDoesntSupportVbnetFiles() =>
-        EmptyCS.Invoking(x => x.AddDocument(@"TestCases\ProjectBuilder.AddDocument.vbhtml").FindDocument("ProjectBuilder.AddDocument.vbhtml").Should().NotBeNull()).Should().Throw<ArgumentException>();
+    public void AddAdditionalDocument_CsharpDoesntSupportVbnetFiles() =>
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/tests/SonarAnalyzer.TestFramework.Test/Build/ProjectBuilderTest.cs:84]:
```suggestion
    public void AddAdditionalDocument_CsharpDoesntSupportVbhtmlFiles() =>
```

File: analyzers/tests/SonarAnalyzer.TestFramework.Test/Build/ProjectBuilderTest.cs [hunk 3]
```diff
@@ -66,27 +66,33 @@ public void AddDocument_InvalidExtension()
...
+        EmptyCS
+            .Invoking(x => x.AddAdditionalDocument(@"TestCases\ProjectBuilder.AddDocument.vbhtml").FindAdditionalDocument("ProjectBuilder.AddDocument.vbhtml").Should().NotBeNull())
+            .Should()
+            .Throw<ArgumentException>();
 
     [TestMethod]
-    public void AddDocument_VbnetDoesntSupportVbnetFiles() =>
-        EmptyVB.Invoking(x => x.AddDocument(@"TestCases\ProjectBuilder.AddDocument.vbhtml").FindDocument("ProjectBuilder.AddDocument.vbhtml").Should().NotBeNull()).Should().Throw<ArgumentException>();
+    public void AddAdditionalDocument_VbnetDoesntSupportVbnetFiles() =>
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/tests/SonarAnalyzer.TestFramework.Test/Build/ProjectBuilderTest.cs:91]:
```suggestion
    public void AddAdditionalDocument_VbnetDoesntSupportVbhtmlFiles() =>
```

File: analyzers/tests/SonarAnalyzer.TestFramework/Build/ProjectBuilder.cs [hunk 4]
```diff
@@ -47,7 +47,10 @@ public Compilation GetCompilation(ParseOptions parseOptions = null, CompilationO
     }
 
     public Document FindDocument(string name) =>
-        project.Documents.Single(d => d.Name == name);
+        project.Documents.SingleOrDefault(d => d.Name == name);
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/tests/SonarAnalyzer.TestFramework/Build/ProjectBuilder.cs:50]: It means that now this method (and the following one) can return `null`, did you check if it's contemplated on the caller side? Do we have test converage for this?

File: analyzers/tests/SonarAnalyzer.TestFramework/Verification/Verifier.cs [hunk 5]
```diff
@@ -269,11 +269,16 @@ private List<string> PrepareRazorFiles(string projectRoot)
 
     private ProjectBuilder CreateProject(bool concurrentAnalysis)
     {
-        var paths = builder.Paths.Select(TestCasePath).ToArray();
+        var paths = builder.Paths.Select(TestCasePath).GroupBy(IsRazorOrCshtml).ToDictionary(x => x.Key, x => x.ToArray());
+        var razorPaths = paths.GetOrAdd(true, _ => []);
+        var documentPaths = paths.GetOrAdd(false, _ => []);
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/tests/SonarAnalyzer.TestFramework/Verification/Verifier.cs:274]: I think we can make it a bit more readable:
```suggestion
var allPaths = builder.Paths.Select(TestCasePath).ToList();
var razorPaths = allPaths.Where(IsRazorOrCshtml).ToArray();
var documentPaths = allPaths.Except(razorPaths).ToArray();
```
not sure performance-wise, but should be negligible anyway, we don't get many paths to go through

Comment (cristian-ambrosini-sonarsource) [reply analyzers/tests/SonarAnalyzer.TestFramework/Verification/Verifier.cs:274]: And what about calling them sourceFilePaths and contentFilesPaths? for now they are just razor/cshtml files but who knows in the future. Let me know wdyt

**Review by cristian-ambrosini-sonarsource — CHANGES_REQUESTED** [review]
Nice! left minor suggestions

File: analyzers/tests/SonarAnalyzer.TestFramework/Build/ProjectBuilder.cs (hunk 4 above)
Comment (sebastien-marichal) [reply analyzers/tests/SonarAnalyzer.TestFramework/Build/ProjectBuilder.cs]: That was legacy that I should have removed 😅

**Review by cristian-ambrosini-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**